## Guided Project: Exploring Hacker News Posts

**Instructions**

Read the hacker_news.csv file in as a list of lists.

Assign the result to the variable hn.

Display the first five rows of hn.


In [30]:
from csv import reader

opened_file = open('hacker_news.csv', encoding = 'utf8')
read_file = reader(opened_file)
hn = list(read_file)
hn_header = hn[0]
hn = hn[1:]

print('csv added successfully')

csv added successfully


**Instructions**

Extract the first row of data, and assign it to the variable headers.

Remove the first row from hn.

Display headers.

Display the first five rows of hn to verify that you removed the header row properly.


In [31]:
print(hn_header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [23]:
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [28]:
print('dataquest' .startswith('Data'))
print('dataquest' .startswith('data'))

False
True


In [29]:
print('Dataquest'.lower())

dataquest


**Instructions**

1. Create three empty lists called ask_posts, show_posts, and other_posts.

2. Loop through each row in hn.
   Assign the title in each row to a variable named title.
   Because the title column is the second column, you'll need to get the element at index 1 in each row.

3. Implement the following steps:
   If the lowercase version of title starts with ask hn, append the row to ask_posts.
   Else if the lowercase version of title starts with show hn, append the row to show_posts.
   Else append to other_posts.

4. Check the number of posts in ask_posts, show_posts, and other_posts.


In [81]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [84]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
 

1744
1162
17194


In [111]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


In [110]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


### Instructions

1. Find the total number of comments in ask posts and assign it to total_ask_comments.
   Set total_ask_comments to 0.

2. Use a for loop to iterate over the ask posts.
   Because the num_comments column is the fifth column in ask_posts, you'll need to get the element at index 4 in each row.
   You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
   Add this value to total_ask_comments.

3. Compute the average number of comments on ask posts and assign it to avg_ask_comments.

4. Print avg_ask_comments.

5. Find the total number of comments in show posts and assign it to total_show_comments.
   Set total_show_comments to 0.

6. Use a for loop to iterate over the show posts.
   Because the num_comments column is the fifth column in show_posts, you'll need to get the element at index 4 in each row.
   You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
   Add this value to total_show_comments.

7. Compute the average number of comments on show posts and assign it to avg_show_comments.

8. Print avg_show_comments.


In [121]:
total_ask_comments = 0

for post in ask_posts:
    total_post = post[4]
    total_post = int( total_post)
    total_ask_comments += total_post
    avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)  

14.038417431192661


In [122]:
total_show_comments = 0

for post in show_posts:
    total_post = post[4]
    total_post = int( total_post)
    total_show_comments += total_post
    avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


### Instructions

1. Import the datetime module as dt.
2. Create an empty list and assign it to result_list. This will be a list of lists.
3. Iterate over ask_posts and append to result_list a list with two elements:
   The first element shall be the column created_at.
   Because the created_at column is the seventh column in ask_posts, you'll need to get the element at index 6 in each row.
   The second element shall be the number of comments of the post.
   You'll also need to convert the value to an integer.
4. Create two empty dictionaries called counts_by_hour and comments_by_hour.
5. Loop through each row of result_list.
6. Extract the hour from the date, which is the first element of the row.
7. Use the datetime.strptime() method to parse the date and create a datetime object.
8. Use the string we want to parse as the first argument and a string that specifies the format as the second argument.
   Use the datetime.strftime() method to select just the hour from the datetime object.
   If the hour isn't a key in counts_by_hour:
   Create the key in counts_by_hour and set it equal to 1.
   Create the key in comments_by_hour and set it equal to the comment number.
   If the hour is already a key in counts_by_hour:
   Increment the value in counts_by_hour by 1.
   Increment the value in comments_by_hour by the comment number.


In [209]:
import datetime as dt
result_list = []
for post in ask_posts:
    created_at = post[6]
    num_of_comments = int(post[4])
    result_list.append([created_at,num_of_comments])
    
count_by_hour = {}
comments_by_hour = {}

for row in result_list:
    extract_time = row[0]
    datetime = dt.datetime.strptime(extract_time,"%m/%d/%Y %H:%M")
    hour = datetime.strftime("%H")
    if hour not in count_by_hour:
        count_by_hour[hour] = 1 
        comments_by_hour[hour] = row[1]
    else:
        count_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
    
    
    

In [210]:
sample_dict = {
    'apple':2,
    'banana':4,
    'orange':6
}
fruits = []
for fruit in sample_dict:
    fruits.append([fruit, 10*sample_dict[fruit]])

In [211]:
fruits

[['apple', 20], ['banana', 40], ['orange', 60]]

**Instructions**

Use the example above to calculate the average number of comments per post for posts created during each hour of the day.

The result should be a list of lists in which the first element is the hour and the second element is the average number of comments per post. Assign the result to a variable named avg_by_hour. Display the results.


In [212]:
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr,comments_by_hour[hr] / count_by_hour[hr]])

In [213]:
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

**Instructions**

1. Create a list that equals avg_by_hour with swapped columns.
   Create an empty list and assign it to swap_avg_by_hour.
   Iterate over the rows of avg_by_hour and append to swap_avg_by_hour a list whose first element is the second element of the row, and whose second element is the first element of the row.
2. Print swap_avg_by_hour.
3. Use the sorted() function to sort swap_avg_by_hour in descending order. Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.
   Set the reverse argument to True, so that the highest value in the first column appears first in the list.
   Assign the result to sorted_swap.
4. Print the string "Top 5 Hours for Ask Posts Comments".
5. Loop through each average and each hour (in this order) in the first five lists of sorted_swap.
6. Use the str.format() method to print the hour and average in the following format: 15:00: 38.59 average comments per post.
   To format the hours, use the datetime.strptime() constructor to return a datetime object and then use the strftime() method to specify the format of the time.
   To format the average, you can use {:.2f} to indicate that just two decimal places should be used.


In [219]:
swap_avg_hour = []

for row in avg_by_hour:
    swap_avg_hour.append([row[1],row[0]])
      

In [220]:
print(swap_avg_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [223]:
sorted_swap = sorted(swap_avg_hour, reverse = True)

In [225]:
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [226]:
print(sorted_swap[:5])

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [238]:
for sort in sorted_swap:
    datetime = dt.datetime.strptime(sort[1],"%H")
    time = datetime.strftime("%H:%M")
    print("{time}:{sort:.2f} average comments per post".format(time = time, sort = sort[0] ))
   

15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post
13:00:14.74 average comments per post
10:00:13.44 average comments per post
14:00:13.23 average comments per post
18:00:13.20 average comments per post
17:00:11.46 average comments per post
01:00:11.38 average comments per post
11:00:11.05 average comments per post
19:00:10.80 average comments per post
08:00:10.25 average comments per post
05:00:10.09 average comments per post
12:00:9.41 average comments per post
06:00:9.02 average comments per post
00:00:8.13 average comments per post
23:00:7.99 average comments per post
07:00:7.85 average comments per post
03:00:7.80 average comments per post
04:00:7.17 average comments per post
22:00:6.75 average comments per post
09:00:5.58 average comments per post


In [ ]:
 time = dt.time(hour = int(row[1])).strftime('%H:%M')
    

In [137]:
datetime[]

datetime.datetime(2016, 5, 21, 9, 22)

In [140]:
datetime.strftime("%H")

'09'

In [142]:
count_by_hour

{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

In [143]:
comments_by_hour

{'09': 90,
 '13': 170,
 '10': 118,
 '14': 214,
 '16': 216,
 '23': 136,
 '12': 146,
 '17': 200,
 '15': 232,
 '21': 218,
 '20': 160,
 '02': 116,
 '18': 218,
 '03': 108,
 '05': 92,
 '19': 220,
 '01': 120,
 '22': 142,
 '08': 96,
 '04': 94,
 '00': 110,
 '06': 88,
 '07': 68,
 '11': 116}